File: clock-backend/pkg/reqres/attendance.go
package reqres

import "github.com/jim/clock-backend/pkg/models"

// CheckInRequest represents the check-in request payload
type CheckInRequest struct {
	Note string `json:"note"`
	// CheckInTime is an optional RFC3339 chosen time to record as the official
	// check-in time. Must be on today's date; the actual press time is always
	// recorded server-side for audit.
	CheckInTime string `json:"check_in_time,omitempty"`
}

// CheckOutRequest represents the check-out request payload
type CheckOutRequest struct {
	Note string `json:"note"`
	// CheckOutTime is an optional RFC3339 chosen time to record as the official
	// check-out time. Must be on today's date and after the check-in time; the
	// actual press time is always recorded server-side for audit.
	CheckOutTime string `json:"check_out_time,omitempty"`
}

// CheckInResponse represents the check-in response
type CheckInResponse struct {
	ID                      int                               `json:"id"`
	EmployeeID              int                               `json:"employee_id"`
	CheckInTime             string                            `json:"check_in_time"`
	Status                  models.AttendanceStatus           `json:"status"`
	Note                    string                            `json:"note,omitempty"`
	PreviousRecordAutoClosed bool                             `json:"previous_record_auto_closed"`
	AutoClosedRecord        *models.AttendanceRecordResponse  `json:"auto_closed_record,omitempty"`
}

// CheckOutResponse represents the check-out response
type CheckOutResponse struct {
	ID           int                     `json:"id"`
	EmployeeID   int                     `json:"employee_id"`
	CheckInTime  string                  `json:"check_in_time"`
	CheckOutTime string                  `json:"check_out_time"`
	Status       models.AttendanceStatus `json:"status"`
	CheckInNote  string                  `json:"check_in_note,omitempty"`
	CheckOutNote string                  `json:"check_out_note,omitempty"`
}

// AttendanceStatusResponse represents the current attendance status
type AttendanceStatusResponse struct {
	IsCheckedIn   bool                              `json:"is_checked_in"`
	CurrentRecord *models.AttendanceRecordResponse  `json:"current_record,omitempty"`
}

// AttendanceRecordsResponse represents the list of attendance records
type AttendanceRecordsResponse struct {
	Records []*models.AttendanceRecordResponse `json:"records"`
	Total   int                                `json:"total"`
}

// SuggestedTimeResponse represents the usual clock time suggestion derived
// from the employee's attendance history
type SuggestedTimeResponse struct {
	Action        string `json:"action"`
	HasSuggestion bool   `json:"has_suggestion"`
	SuggestedTime string `json:"suggested_time,omitempty"`
	SampleCount   int    `json:"sample_count"`
}

// ErrorResponse represents an error response
type ErrorResponse struct {
	Error string `json:"error"`
}

File: clock-frontend/types/attendance.ts
export type AttendanceStatus = 'CHECKED_IN' | 'CHECKED_OUT' | 'AUTO_CLOSED';

export interface AttendanceRecord {
  id: number;
  employee_id: number;
  check_in_time: string;
  check_out_time: string | null;
  actual_check_in_time?: string | null;
  actual_check_out_time?: string | null;
  status: AttendanceStatus;
  check_in_note: string | null;
  check_out_note: string | null;
}

export interface CheckInRequest {
  note?: string;
  // Optional chosen time (RFC3339) recorded as the official check-in time;
  // must be on today's date. The actual press time is stored server-side.
  check_in_time?: string;
}

export interface CheckInResponse {
  id: number;
  employee_id: number;
  check_in_time: string;
  status: AttendanceStatus;
  note: string | null;
  previous_record_auto_closed: boolean;
  auto_closed_record?: {
    id: number;
    check_in_time: string;
    status: AttendanceStatus;
  };
}

export interface CheckOutRequest {
  note?: string;
  // Optional chosen time (RFC3339) recorded as the official check-out time;
  // must be on today's date and after check-in. The actual press time is
  // stored server-side.
  check_out_time?: string;
}

export interface CheckOutResponse extends AttendanceRecord {}

export type ClockAction = 'check_in' | 'check_out';

export interface SuggestedTimeResponse {
  action: ClockAction;
  has_suggestion: boolean;
  suggested_time?: string;
  sample_count: number;
}

export interface AttendanceStatusResponse {
  is_checked_in: boolean;
  current_record: AttendanceRecord | null;
}

export interface AttendanceRecordsResponse {
  records: AttendanceRecord[];
  total: number;
}

export type EditRequestStatus = 'PENDING' | 'APPROVED' | 'REJECTED';

export type EditRequestType = 'EDIT' | 'ADD';

export interface EditRequest {
  id: number;
  attendance_record_id: number | null;
  employee_id: number;
  request_type: EditRequestType;
  requested_check_in_time: string | null;
  requested_check_out_time: string | null;
  note: string | null;
  status: EditRequestStatus;
  reviewed_by: number | null;
  reviewed_at: string | null;
  review_note: string | null;
  created_at: string;
}

export interface SubmitEditRequestPayload {
  attendance_record_id: number;
  check_in_time?: string;
  check_out_time?: string;
  note: string;
}

export interface SubmitAddRequestPayload {
  check_in_time: string;
  check_out_time?: string;
  note: string;
}

export interface EditRequestListResponse {
  edit_requests: EditRequest[];
  total: number;
}
